### PR TITLE
[Merged by Bors] - Expand eth1 block cache, add more logs

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

- Caches later blocks than is required by `ETH1_FOLLOW_DISTANCE`.
- Adds logging to `warn` if the eth1 cache is insufficiently primed.
- Use `max_by_key` instead of `max_by` in `BeaconChain::Eth1Chain` since it's simpler.
- Rename `voting_period_start_timestamp` to `voting_target_timestamp` for accuracy.

## Additional Info

The reason for eating into the `ETH1_FOLLOW_DISTANCE` and caching blocks that are closer to the head is due to possibility for `SECONDS_PER_ETH1_BLOCK` to be incorrect (as is the case for the Pyrmont testnet on Goerli).

If `SECONDS_PER_ETH1_BLOCK` is too short, we'll skip back too far from the head and skip over blocks that would be valid [`is_candidate_block`](https://github.com/ethereum/eth2.0-specs/blob/v1.0.0/specs/phase0/validator.md#eth1-data) blocks. This was the case on the Pyrmont testnet and resulted in Lighthouse choosing blocks that were about 30 minutes older than is ideal.

### CODE REVIEW

File: beacon_node/eth1/src/service.rs
Comment (michaelsproul) [inline beacon_node/eth1/src/service.rs:192]:
```suggestion
    /// Returns the follow distance that has been shortened to accommodate for differences in the
```

File: book/src/api-lighthouse.md
Comment (michaelsproul) [inline book/src/api-lighthouse.md:195]: "latest" seems ambiguous here. We mean "oldest" don't we? Not "most recent"?

File: beacon_node/eth1/src/service.rs
Comment (michaelsproul) [inline beacon_node/eth1/src/service.rs:127]:
```suggestion
        // Compute the number of extra blocks we store prior to the voting period start blocks.
```

**Review by michaelsproul — APPROVED** [review]
LGTM modulo typo fixes, although I would also like an Eth1 expert (@pawanjay176) to review this, either before or after merging.

File: book/src/api-lighthouse.md
Comment (paulhauner) [reply book/src/api-lighthouse.md:195]: We do mean most recent.

In an eth1 voting period we are permitted to vote on a range of timestamps. This is the highest timestamp in that range.

File: beacon_node/eth1/src/service.rs
Comment (pawanjay176) [inline beacon_node/eth1/src/service.rs:927]: I'm not very sure what this test is checking here. `block_cache_truncation` could be less than the old block cache size and the test would still pass.

**Review by pawanjay176 — APPROVED** [review]
LGTM! 
Just a nit and a question about the test. 
I think we should eventually refactor the block cache such that it truncates based on timestamps instead of blocks, but this works well too for the time being :)

File: beacon_node/beacon_chain/src/eth1_chain.rs
Comment (paulhauner) [reply beacon_node/beacon_chain/src/eth1_chain.rs:453]: I think it is the block number. It's the result of the `get_votes_to_consider` function which is `(Eth1Data, BlockNumber)`. Perhaps you're confusing it with the output of `collect_valid_votes`?

File: beacon_node/eth1/src/service.rs
Comment (paulhauner) [reply beacon_node/eth1/src/service.rs]: Basically I'm trying to check that with all our fancy math we don't end up with something smaller than the very minimum required. I added some more complexity to it to make it a better test :) Thanks.

**Comment by paulhauner** [comment]
> I think we should eventually refactor the block cache such that it truncates based on timestamps instead of blocks, but this works well too for the time being :)

I agree, the whole `eth1` crate is in a need of a refactor at this point. We can do this post-mainnet :)

**Pull request closed without merging**
